app/services/vot/acr_component_values.rb

```diff
@@ -1,105 +1,306 @@
       implied_component_values: [],
       requirements: [:aal2],
     ).freeze
-    AAL2_PHISHING_RESISTANT = ComponentValue.new(
+    AAL2_PHISHING_RESISTANT = AcrComponentValue.new(
       name: Saml::Idp::Constants::AAL2_PHISHING_RESISTANT_AUTHN_CONTEXT_CLASSREF,
       description: 'Legacy AAL2 with phishing resistance',
       implied_component_values: [],
       requirements: [:aal2, :phishing_resistant],
     ).freeze
-    AAL2_HSPD12 = ComponentValue.new(
+    AAL2_HSPD12 = AcrComponentValue.new(
       name: Saml::Idp::Constants::AAL2_HSPD12_AUTHN_CONTEXT_CLASSREF,
       description: 'Legacy AAL2 with HSPD12',
       implied_component_values: [],
       requirements: [:aal2, :hspd12],
     ).freeze
-    AAL3 = ComponentValue.new(
+    AAL3 = AcrComponentValue.new(
       name: Saml::Idp::Constants::AAL3_AUTHN_CONTEXT_CLASSREF,
       description: 'Legacy AAL3',
       implied_component_values: [],
       requirements: [:aal2, :phishing_resistant],
     ).freeze
-    AAL3_HSPD12 = ComponentValue.new(
+    AAL3_HSPD12 = AcrComponentValue.new(
       name: Saml::Idp::Constants::AAL3_HSPD12_AUTHN_CONTEXT_CLASSREF,
       description: 'Legacy AAL3 with HSPD12',
       implied_component_values: [],
       requirements: [:aal2, :hspd12],
     ).freeze
 
-    NAME_HASH = constants.map do |constant|
-      component_value = const_get(constant)
-      [component_value.name, component_value]
-    end.to_h.freeze
+    # @type [Hash]
+    NAME_HASH = constants(false).
+      map { |c| const_get(c, false) }.
+      filter { |c| c.is_a?(AcrComponentValue) }.
+      index_by(&:name).freeze
+
+    VALUES = NAME_HASH.values.freeze
+
+    DELIM = ' '
+
+    IAL_COMPONENTS_BY_NAME_HASH = [
+      LOA1,
+      LOA3,
+      IAL1,
+      IALMAX,
+      IAL2,
+      IAL2_BIO_PREFERRED,
+      IAL2_BIO_REQUIRED,
+    ].index_by(&:name).freeze
+
+    # IAL components ordered by priority, descending
+    IAL_COMPONENTS_BY_PRIORITY_LIST = [
+      IAL2_BIO_REQUIRED,
+      IAL2_BIO_PREFERRED,
+      IAL2,
+      LOA3,
+      IALMAX,
+      IAL1,
+      LOA1,
+    ].freeze
+
+    # IAL names ordered by priority, descending
+    IALS_BY_PRIORITY_LIST = IAL_COMPONENTS_BY_PRIORITY_LIST.map(&:name).freeze
+
+    # AAL components ordered by priority, descending
+    AAL_COMPONENTS_BY_NAME_HASH = [
+      DEFAULT_AAL,
+      AAL1,
+      AAL2,
+      AAL2_PHISHING_RESISTANT,
+      AAL2_HSPD12,
+      AAL3,
+      AAL3_HSPD12,
+    ].index_by(&:name).freeze
+
+    # AAL components ordered by priority, descending
+    AAL_COMPONENTS_BY_PRIORITY_LIST = [
+      AAL2_HSPD12,
+      AAL3_HSPD12,
+      AAL2_PHISHING_RESISTANT,
+      AAL3,
+      AAL2,
+      AAL1,
+      DEFAULT_AAL,
+    ].freeze
+
+    # AAL names ordered by priority, descending
+    AALS_BY_PRIORITY_LIST = AAL_COMPONENTS_BY_PRIORITY_LIST.map(&:name).freeze
+
+    # All ACR components ordered by type (IAL, AAL) and then descending priority
+    ACR_COMPONENTS_BY_PRIORITY_LIST = [
```

Comment:
[question] hm i am not sure it makes sense to smush IAL and AAL in a priority list -- that implies that IAL1 is a stricter priority than, say, phishing resistance, when they're really two separate concerns. is there prior art you're copying, or a reason i'm overlooking for doing it this way?